src/Translation_unit.mli

```diff
@@ -33,34 +32,34 @@ type 'a t =
   ; normalize: Conf.t -> 'a with_comments -> 'a
   ; printast: Caml.Format.formatter -> 'a -> unit }
 
-(** Existential package of a type of translation unit and its operations. *)
-type x = XUnit : 'a t -> x
-
-exception Warning50 of (Location.t * Warnings.t) list
-
-type result =
-  | Ok
-  | Invalid_source of exn
-  | Unstable of int
-  | Ocamlformat_bug of exn
+type error =
+  | Invalid_source of {input_name: string; exn: exn}
+  | Unstable of
+      { input_name: string
+      ; iteration: int
+      ; prev: string
+      ; next: string }
+  | Ocamlformat_bug of {input_name: string; exn: exn}
   | User_error of string
 
-val parse :
-     (Lexing.lexbuf -> 'a)
-  -> Conf.t
-  -> input_file:string
-  -> In_channel.t
-  -> 'a with_comments
+val parse : (Lexing.lexbuf -> 'a) -> Conf.t -> string -> 'a with_comments
```

Comment:
`format` takes a `('a with_comments, exn) result` so that it can be used by both ocamlformat and ocamlformat_reason. `parse` is used by `ocamlformat.ml` to generate that `('a with_comments, exn) result` value. I don't see a better solution/api at the moment.

Reply:
Yes, ok. I guess that if integrated e.g. into the ppx pipeline of dune, we'd not use the special treatment of warnings and hash-bang lines in `parse`, and just pass the parsetree that comes down the pipe to `format`.